Davidvc conceptual model (#140)
pact-foundation/docs.pact.io: website/docs/getting_started/conceptual_overview.md

# Conceptual Overview 

Using Pact successfully requires you to understand how Pact works and the concepts and terminology behind Pact. 
If you try to build your CI/CD without this foundational understanding, you may find yourself getting confused or getting unexpected results.

## Consumers and publishing pacts

Whenever a consumer runs a contract test for a particular provider and it succeeds, a file is generated that records 
all of the interactions between the consumer and the provider. This record is called a pact.

In your CI/CD environment, every time you run the contract tests, the resulting pact will be published to the Pact Broker. 
This is called a pact publication.

Whenever the consumer changes, this is called a __consumer version__. The version is often identified by its git hash. 
You may also want to add other distinguishing information, such as the feature toggle name if you are using feature 

> toggles.  In general, you want to uniquely identify each version of the consumer that may be deployed to an environment which may require a different contract.

Author: maybe add _or feature branches, if you are using feature branches._

and we can provide further reading when we have conceptual overviews for those

- Multiple versions of the consumer can have the same pact version (if the changes to the consumer don't change the pact)
- A specific version of the consumer will always have exactly one pact version for each provider.

## Branches & Environments

:::info THE GOLDEN RULE FOR LABELLING
Label with the branch name when you publish pacts or verification results, and record deployment with the environment name after you deploy.

- [branches](/pact_broker/branches)
- [recording deployments and releases](/pact_broker/recording_deployments_and_releases)

::: 

At any point you may want to label a particular version of the consumer with one or more identifiers. For example, 
you can identify the consumer with its current branch (e.g. a PR branch or the main branch), or you can identify it with the 
environment it has been deployed to.

Branches in the Pact Broker are designed to model repository (git, svn etc) branches. A branch in the Pact Broker belongs to a pacticipant (application). A branch may have many pacticipant versions, and a pacticipant version may belong to many branches (but typically, it will belong to just one).

> Remember that a pacticipant version in the Pact Broker should map 1:1 to a commit in your repository. To facilitate this, the version number used to publish pacts and verification results should either be or contain the commit.

Reviewer: This isn't true if you are using feature toggles or other configuration systems that change the behavior depending on a configuration. The same commit in a repo may interact with different versions of a pact, and therefore require different pacticipant versions.

Reviewer: Also, you mention the concept of "pacticipant" here but this concept isn't introduced until the end of the article

Author: Ahh so this comment was lifted from here 

https://docs.pact.io/pact_broker/branches#domain-model

I think the feature branches which is being demonstrated here, is distinctly different to the challenges of feature toggles, and I wonder if that changes alongside whether they can be turned on pre deploy, or during run time, with something like unleash or reading env vars from SSM.

Author: and it is a good shout on the term pacticipant as we introduce that in the next section

> We recommend that you set the branch property when you publish pacts and verification results, and use record-deployment or record-release when you deploy/release.

Author: this line and the one below it feel like at odds, maybe we don't need the one above.

I feel the We recommend in stronger but the general recommendation suggests more of a community consensus rather than dictatorial

Reviewer: I agree probably want to get rid of "should"

For example, let's say you have a PetStoreConsumer application.  Every time you change it that's a new version 
of the application.  When you run contract tests for that version, you publish a new pact publication to the broker, 
which may or may not generate a new pact version.

| Date  | Consumer Versions | branches     | environment | Pact Version |
| ----- | ----------------- | ------------ | ----------- | ------------ |
| 12/21 | 63be8b            | prod, main   | production  | 1            |
| 12/22 | 9ecb61            | staging,main | staging     | 2            |
| 12/23 | 31696e            | feat-xyz     | N/A         | 2            |
| 12/23 | 3b4de7            | feat-abc     | N/A         | 3            |

Here we see that:
- Version 63be8b is deployed to prod and is in the main branch.
- Version 9ecb61 is deployed to staging, and is also in the main branch. It has changed the pact, so it has a new pact version.
- Version 31696e is in a feature branch and has not been deployed anywhere. It did not change the pact, so the pact version remains the same
- Version 3b4de7 is in a different feature branch. It did change the pact, so we now have a new pact version
  

> :::warning A caveat about tags

Author: we now haven't mentioned tags above, so not sure if this should be brought up.

or should it something like, familiar with Pact and using tags, see the below notice/warning.

![Provider domain model](../media/conceptual_overview_images/provider-domain-model.png)

Let's expand the example from above to include the provider and its verifications.

| Date  | Provider Version | branches     | environment | Verified against pact versions |
| ----- | ----------------- | ------------ | ----------- | ------------ |
| 12/21 | 6464f9            | prod, main   | production  | 1,2            |
| 12/22 | a8513e            | staging,main | staging     | 1,2,3           |
| 12/23 | 35a850            | feat-xyz     | N/A         | 2,3            |
| 12/23 | 741a54            | feat-abc     | N/A         | 3            |

Note that provider version 35a850 in the feat-X branch has not been verified against pact version 1. Because that's the 
version that the prod consumer is at, it is not safe to deploy this version of the provider to prod. You probably 
also don't want to merge it into main until this gets fixed. Similarly provider version 741a54 has only verified 
pact version 3. So it's not safe to deploy to either staging or prod, and should definitely not be merged to main.

## Generalizing: the pacticipant

> The Pact team has generalized the "consumer" and "provider" roles of a pact using the term __pacticipant__. A particular 

Reviewer: ```suggestion
We have generalized the "consumer" and "provider" roles of a pact using the term __pacticipant__. A particular 
```
